src/main/java/net/glowstone/command/minecraft/TestForBlockCommand.java

```diff
@@ -0,0 +1,117 @@
+package net.glowstone.command.minecraft;
+
+import net.glowstone.block.GlowBlock;
+import net.glowstone.block.state.BlockStateData;
+import net.glowstone.block.state.InvalidBlockStateException;
+import net.glowstone.block.state.StateSerialization;
+import net.glowstone.command.CommandUtils;
+import net.glowstone.constants.ItemIds;
+import net.glowstone.util.mojangson.Mojangson;
+import net.glowstone.util.mojangson.ex.MojangsonParseException;
+import net.glowstone.util.nbt.CompoundTag;
+import org.bukkit.ChatColor;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.defaults.VanillaCommand;
+import org.bukkit.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class TestForBlockCommand extends VanillaCommand {
+    public TestForBlockCommand() {
+        super("testforblock",
+                "Tests for a certain block at a given location",
+                "/testforblock <x> <y> <z> <block> [dataValue|state] [dataTag]",
+                Collections.emptyList());
+        setPermission("minecraft.command.testforblock");
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String label, String[] args) {
+        if (!testPermission(sender)) {
+            return false;
+        }
+        if (args.length < 4) {
+            sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
+            return false;
+        }
+        String itemName = args[3];
+        if (!itemName.startsWith("minecraft:")) {
+            itemName = "minecraft:" + itemName;
+        }
+        Material type = ItemIds.getItem(itemName);
+        Location location = CommandUtils.getLocation(CommandUtils.getLocation(sender), args[0], args[1], args[2]);
+        GlowBlock block = (GlowBlock) location.getBlock();
+        if (block.getType() != type) {
+            sender.sendMessage(ChatColor.RED + "The block at " + location.getBlockX() + ", " + location.getBlockY() + ", " + location.getBlockZ() +
+                    " is " + ItemIds.getName(block.getType()) + " (expected: " + ItemIds.getName(type) + ")");
+            return false;
+        }
+        if (args.length > 4) {
+            String state = args[4];
```

Comment:
Since we know this same "check the data value or state argument" is used in at least three commands (clone and testforblock, and testforblocks), mind abstracting this check out into a CommandUtils method? May want to return a custom type with two properties, `Integer data` and `BlockStateData stateData`, for ease of use in these commands.

Reply:
Will do.